private/pkg/oauth2/client.go

```diff
@@ -0,0 +1,191 @@
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, resp.Body.Close())
+	}()
+
+	payload := &struct {
+		Error
+		DeviceAuthorizationResponse
+	}{}
+	if err := c.parseJSONResponse(resp, payload); err != nil {
+		return nil, fmt.Errorf("oauth2: invalid response: %w", err)
+	}
+	if payload.ErrorCode != "" {
+		return nil, &payload.Error
+	}
+	if code := resp.StatusCode; code != http.StatusOK {
+		return nil, fmt.Errorf("oauth2: invalid status: %v", code)
+	}
+	return &payload.DeviceAuthorizationResponse, nil
+}
+
+// AccessDeviceToken polls the authorization server for the device access token. The interval
+// parameter specifies the polling interval in seconds.
+func (c *Client) AccessDeviceToken(ctx context.Context, interval int, args *DeviceAccessTokenRequest) (*DeviceAccessTokenResponse, error) {
+	encodedValues := args.ToValues().Encode()
+	if interval <= 0 {
+		interval = 5 // Default polling interval of 5 seconds.
+	} else if interval > 30 {
+		return nil, fmt.Errorf("oauth2: interval is too large: %v", interval)
+	}
+	ticker := time.NewTicker(time.Duration(interval) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-ticker.C:
+			req, err := http.NewRequestWithContext(ctx, "POST", c.baseURL+DeviceTokenPath, strings.NewReader(encodedValues))
+			if err != nil {
+				return nil, err
+			}
+			req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+			req.Header.Set("Accept", "application/json")
+
+			rsp, err := c.client.Do(req)
+			if err != nil {
+				return nil, err
+			}
+			payload := &struct {
+				Error
+				DeviceAccessTokenResponse
+			}{}
+			if err := c.parseJSONResponse(rsp, payload); err != nil {
+				if closeErr := rsp.Body.Close(); closeErr != nil {
+					err = multierr.Append(err, closeErr)
+				}
+				return nil, fmt.Errorf("oauth2: invalid response: %w", err)
+			}
+			if err := rsp.Body.Close(); err != nil {
+				return nil, fmt.Errorf("oauth2: failed to close response body: %w", err)
+			}
+			if rsp.StatusCode == http.StatusOK && payload.ErrorCode == "" {
+				return &payload.DeviceAccessTokenResponse, nil
+			}
+			switch payload.ErrorCode {
+			case ErrorCodeSlowDown:
+				// If the server is rate limiting the client, increase the polling interval.
+				interval += 5
+				ticker.Reset(time.Duration(interval) * time.Second)
+			case ErrorCodeAuthorizationPending:
+				// If the user has not yet authorized the device, continue polling.
+				continue
+			case ErrorCodeAccessDenied, ErrorCodeExpiredToken:
+				// If the user has denied the device or the token has expired, return the error.
+				return nil, &payload.Error
+			default:
+				return nil, &payload.Error
+			}
+		}
+	}
+}
+
+func (c *Client) parseJSONResponse(rsp *http.Response, payload any) error {
```

Comment:
`rsp, resp, etc` should be `response`, same comment for request-like things should be `request` and not `req, args, etc`